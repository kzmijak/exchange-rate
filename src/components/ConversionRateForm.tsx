import { Button, FormControl, FormLabel, Input, Stack } from "@mui/joy";
import { observer } from "mobx-react-lite";
import { useRootStore } from "modules/RootStore";
import { ChangeEvent, FC, FormEvent, useState } from "react";
import { NumericFormat } from "react-number-format";

type ConversionRateFormProps = {
  onSubmitSuccess: () => void;
};
export const ConversionRateForm: FC<ConversionRateFormProps> = observer(
  ({ onSubmitSuccess }) => {
    const { conversionRateStore } = useRootStore();
    const [inputRate, setInputRate] = useState(
      conversionRateStore.eurToPln.toString()
    );

    const handleRateInputChange = (event: ChangeEvent<HTMLInputElement>) => {
      setInputRate(event.target.value);
    };

    const handleSubmit = (event: FormEvent) => {
      event.preventDefault();
      if (inputRate) {
        const rateNumber = Number(inputRate);
        if (!isNaN(rateNumber)) {
          conversionRateStore.changeRate(rateNumber);
          onSubmitSuccess();
        }
      }
    };

    return (
      <Stack component="form" onSubmit={handleSubmit} spacing={3}>
        <FormControl>
          <FormLabel>EUR to PLN</FormLabel>
          <NumericFormat
            customInput={Input}
            autoFocus
            placeholder="Ex. 4.382"
            onChange={handleRateInputChange}
            value={conversionRateStore.eurToPln}
          />
        </FormControl>
        <Button type="submit">Apply</Button>
      </Stack>
    );
  }
);
