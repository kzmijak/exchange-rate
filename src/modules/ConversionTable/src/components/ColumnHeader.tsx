import { Box, Typography } from "@mui/joy";
import { FC } from "react";
import { columnWidths } from "../config/columnWidths";
import { ConversionTableColumn } from "../models/ConversionTableColumn";
import { columnHeaderNames } from "../config/columnHeaderNames";

type ColumnHeaderProps = {
  column: ConversionTableColumn;
};

export const ColumnHeader: FC<ColumnHeaderProps> = ({ column }) => {
  return (
    <Box component="th" width={columnWidths[column]}>
      <Typography>{columnHeaderNames[column]}</Typography>
    </Box>
  );
};
