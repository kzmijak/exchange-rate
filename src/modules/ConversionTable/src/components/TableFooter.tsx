import { FC } from "react";
import { ConversionTableRow } from "../models/ConversionTableRow";
import { Typography } from "@mui/joy";
import { convertPlnToEur } from "utils/convertCurrency";

type TableFooterProps = {
  rows: ConversionTableRow[];
};

export const TableFooter: FC<TableFooterProps> = ({ rows }) => {
  const sumPLN = rows.reduce((acc, curr) => acc + curr.amountPLN, 0);
  const sumEUR = convertPlnToEur(sumPLN);

  return (
    <tfoot>
      <tr>
        <Typography level="h6" textAlign="left">
          Sum: {sumPLN} PLN ( {sumEUR} EUR )
        </Typography>
      </tr>
    </tfoot>
  );
};
