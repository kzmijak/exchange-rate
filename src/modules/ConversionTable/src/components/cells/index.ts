export { TitleCell } from "./TitleCell";
export { AmountEURCell } from "./AmountEURCell";
export { AmountPLNCell } from "./AmountPLNCell";
export { OptionsCell } from "./OptionsCell";
