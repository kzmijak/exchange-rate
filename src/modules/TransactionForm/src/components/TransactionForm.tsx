import { FC } from "react";
import { TransactionFormContent } from "../models/TransactionFormContent";
import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import { schema } from "../utils/schema";
import { Button, Input, Stack, Typography } from "@mui/joy";

type TransactionFormProps = {
  defaultValues?: Partial<TransactionFormContent>;
  onSubmit: (formContent: TransactionFormContent) => void;
};

export const TransactionForm: FC<TransactionFormProps> = ({
  defaultValues,
  onSubmit,
}) => {
  const { register, handleSubmit } = useForm<TransactionFormContent>({
    defaultValues,
    resolver: yupResolver(schema),
  });

  return (
    <Stack component="form" onSubmit={handleSubmit(onSubmit)}>
      <Stack flex={1}>
        <Stack direction="row">
          <Typography>Title of transaction</Typography>
          <Input {...register("title")} />
        </Stack>
        <Stack direction="row">
          <Typography>Amount (in PLN)</Typography>
          <Input {...register("amount")} />
        </Stack>
      </Stack>
      <Button type="submit">Add</Button>
    </Stack>
  );
};
