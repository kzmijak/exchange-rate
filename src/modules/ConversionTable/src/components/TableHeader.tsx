import { FC } from "react";
import { conversionTableColumnConsts } from "../models/ConversionTableColumn";
import { ColumnHeader } from "./ColumnHeader";

export const TableHeader: FC = () => {
  return (
    <thead>
      <tr>
        {conversionTableColumnConsts.map((column) => (
          <ColumnHeader key={column} column={column} />
        ))}
      </tr>
    </thead>
  );
};
