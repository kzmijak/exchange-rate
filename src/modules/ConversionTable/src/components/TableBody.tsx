import { FC } from "react";
import { ConversionTableRow } from "../models/ConversionTableRow";
import { AmountEURCell, AmountPLNCell, OptionsCell, TitleCell } from "./cells";

export type TableBodyProps = {
  rows: ConversionTableRow[];
};

export const TableBody: FC<TableBodyProps> = ({ rows }) => {
  return (
    <tbody>
      {rows.map((row) => (
        <tr key={row.id} style={{ cursor: "pointer" }}>
          <TitleCell row={row} />
          <AmountPLNCell row={row} />
          <AmountEURCell row={row} />
          <OptionsCell row={row} />
        </tr>
      ))}
    </tbody>
  );
};
