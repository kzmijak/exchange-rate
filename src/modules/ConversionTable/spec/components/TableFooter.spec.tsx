import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { TableFooter } from "../../src/components/TableFooter";
import { rows } from "../__fixtures__/rows";

describe("TableFooter", () => {
  const Component = () => <TableFooter rows={rows} />;

  it("should match snapshot", () => {
    expect(render(<Component />).container).toMatchSnapshot();
  });

  it("should properly calculate sums", () => {
    render(<Component />);

    expect(screen.getByText(/Sum:\s*122.55\s*PLN\s*\(\s*27.97\s*EUR\s*\)/)); // Sum: 122.55 PLN (27.97 EUR)
  });
});
