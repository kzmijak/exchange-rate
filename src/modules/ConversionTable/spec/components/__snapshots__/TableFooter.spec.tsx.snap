// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableFooter > should match snapshot 1`] = `
<div>
  <tfoot>
    <tr>
      <h6
        class="MuiTypography-root MuiTypography-h6 css-m69daf-JoyTypography-root"
      >
        Sum: 
        122.55
         PLN ( 
        27.97
         EUR )
      </h6>
    </tr>
  </tfoot>
</div>
`;
